src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp

```diff
@@ -51,16 +51,106 @@ FileSegmentRangeWriter::FileSegmentRangeWriter(
 {
 }
 
-bool FileSegmentRangeWriter::write(const char * data, size_t size, size_t offset, bool is_persistent)
+bool FileSegmentRangeWriter::write(const char * data, size_t size, size_t offset, FileSegmentKind segment_kind)
+{
+    size_t written_size = writeImpl(data, size, offset, segment_kind, true);
+    return written_size == size;
+}
+
+size_t FileSegmentRangeWriter::tryWrite(const char * data, size_t size, size_t offset, FileSegmentKind segment_kind)
+{
+    return writeImpl(data, size, offset, segment_kind, false);
+}
+
+size_t FileSegmentRangeWriter::writeImpl(const char * data, size_t total_size, size_t offset, FileSegmentKind segment_kind, bool strict)
+{
+    size_t total_written_size = 0;
+    while (total_size > 0)
+    {
+        size_t reserved_size = reserveInCurrentSegment(total_size, offset, segment_kind, strict);
+        chassert(reserved_size <= total_size);
+
+        writeCurrentSegment(data, reserved_size, offset);
+        if (reserved_size == 0)
+            break;
+
+        data += reserved_size;
+
+        total_size -= reserved_size;
+        offset += reserved_size;
+        total_written_size += reserved_size;
+    }
+    return total_written_size;
+}
+
+size_t FileSegmentRangeWriter::reserveImpl(size_t total_size, size_t offset, FileSegmentKind segment_kind, bool strict)
+{
+    size_t total_reserved_size = 0;
+    while (total_size > 0)
+    {
+        size_t reserved_size = reserveInCurrentSegment(total_size, offset, segment_kind, strict);
+        chassert(reserved_size <= total_size);
+        if (reserved_size == 0)
+            break;
+
+        auto file_segment = *current_file_segment_it;
+        file_segment->setDownloadedSize(reserved_size);
```

Comment:
Maybe we expose some method `FileCache::getOrSet(const Key & key, size_t size, FileSegmentKind);` that would allocate a much space as needed and return filename there to write data and it's guaranteed that we have enough space to write `size` amount of bytes? And if we call it with different `size` it would shrink reserved space or try to extend it?

Reply:
> So, it's impossible to reserve space without writing any data though the FileSegment?

Yes, because there are assertions that check file size on the filesystem https://github.com/ClickHouse/ClickHouse/blob/52c13559bd8da39918bf232e7de6ae68a63d6e83/src/Interpreters/Cache/FileSegment.cpp#L670
and logical errors:
https://github.com/ClickHouse/ClickHouse/blob/52c13559bd8da39918bf232e7de6ae68a63d6e83/src/Interpreters/Cache/FileCache.cpp#L139-L153

> Can we create files bigger than segment size (final temporary file can have arbitrary size)? Or should we spit them in this case?

No. This is why `FIleSegmentRangeWriter` was added - for the case when you want to write data with unknown result size (used for caching data on INSERT queries, and as you have seen - it allocates file segment of max_file_segment size and splits the data between them automatically )